include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <variant>
+
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+class MatchAction {
+public:
+  static MatchAction skip() { return MatchAction(absl::variant<bool, std::string>(true)); }
+
+  static MatchAction callback(std::string callback) {
+    return MatchAction(absl::variant<bool, std::string>(std::move(callback)));
+  }
+
+  static MatchAction fromProto(envoy::config::common::matcher::v3::MatchTree::MatchAction proto) {
+    if (proto.skip()) {
+      return skip();
+    } else {
+      return callback(proto.callback());
+    }
+  }
+
+  absl::optional<absl::string_view> callback() {
+    if (absl::holds_alternative<std::string>(action_)) {
+      return absl::string_view(std::get<std::string>(action_));
+    }
+
+    return absl::nullopt;
+  }
+
+  bool isSkip() { return std::get<bool>(action_); }
+
+private:
+  explicit MatchAction(absl::variant<bool, std::string> action) : action_(action) {}
+  absl::variant<bool, std::string> action_;
```

Comment:
nit: prefer dual state enum for clarity vs. bool or just an empty struct which acts as a named placeholder for skip.

Reply:
Also, nit: setup a using alias for this variant type.